### DISCUSSION TITLE
shap.DeepExplainer error with tensorflow 2.8.0

### DISCUSSION THREAD

**Discussion by yerbby** [shap/shap#2442] [Q&A]
Hi, 

the standard test code: 

```
import shap
import numpy as np

# select a set of background examples to take an expectation over
background = x_train[np.random.choice(x_train.shape[0], 100, replace=False)]

# explain predictions of the model on three images
e = shap.DeepExplainer(model, background)
# ...or pass tensors directly
# e = shap.DeepExplainer((model.layers[0].input, model.layers[-1].output), background)
shap_values = e.shap_values(x_test[1:5])

```
rises an error with latest tensorflow.keras versions (at least in 2.8.x), as below. 

Any hints are highly appreciated.
Best regards.

----------------------------------------------------------------------------------------------------------------------
keras is no longer supported, please use tf.keras instead.
Your TensorFlow version is newer than 2.4.0 and so graph support has been removed in eager mode and some static graphs may not be supported. See PR #1483 for discussion.
`tf.keras.backend.set_learning_phase` is deprecated and will be removed after 2020-10-11. To update it, simply pass a True/False value to the `training` argument of the `__call__` method of your layer or model.
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
[<ipython-input-11-efcedd8eb358>](https://localhost:8080/#) in <module>()
      8 # ...or pass tensors directly
      9 # e = shap.DeepExplainer((model.layers[0].input, model.layers[-1].output), background)
---> 10 shap_values = e.shap_values(test_dataset[1:5])

6 frames
[/usr/local/lib/python3.7/dist-packages/tensorflow/python/framework/func_graph.py](https://localhost:8080/#) in autograph_handler(*args, **kwargs)
   1145           except Exception as e:  # pylint:disable=broad-except
   1146             if hasattr(e, "ag_error_metadata"):
-> 1147               raise e.ag_error_metadata.to_exception(e)
   1148             else:
   1149               raise

AttributeError: in user code:

    File "/usr/local/lib/python3.7/dist-packages/shap/explainers/_deep/deep_tf.py", line 243, in grad_graph  *
        out = self.model(shap_rAnD)
    File "/usr/local/lib/python3.7/dist-packages/keras/utils/traceback_utils.py", line 67, in error_handler  **
        raise e.with_traceback(filtered_tb) from None
    File "/usr/local/lib/python3.7/dist-packages/shap/explainers/_deep/deep_tf.py", line 26, in custom_record_gradient
        out = tf_backprop._record_gradient("shap_"+op_name, inputs, attrs, results)

    AttributeError: Exception encountered when calling layer "tf.__operators__.getitem_4" (type SlicingOpLambda).
    
    module 'tensorflow.python.eager.backprop' has no attribute '_record_gradient'
    
    Call arguments received:
      • tensor=tf.Tensor(shape=(200, 224, 224, 3), dtype=float32)
      • slice_spec=['Ellipsis', {'start': 'None', 'stop': 'None', 'step': '-1'}]
      • var=None

**Comment by nathan-deepmm** [comment]
Any updates? Is DeepExplainer supported for TF 2.x?

**Reply by fesaaved** [reply]
I solved similar issue adding the line:
`tf.compat.v1.disable_v2_behavior()`
The solution was found here: 
https://stackoverflow.com/questions/66814523/shap-deepexplainer-with-tensorflow-2-4-error

**Reply by ericballabene** [reply]
raise NotImplementedError(
NotImplementedError: numpy() is only available when eager execution is enabled.

from     model = tf.keras.models.load_model(mymodel)

**Reply by fesaaved** [reply]
@ericballabene you are right. 
What I did is to train my model without using `tf.compat.v1.disable_v2_behavior()` save the model and then in a second routine apply shap values using `tf.compat.v1.disable_v2_behavior()`. However, the model predictions change a lot when we disable the v2 behavior, so I think this is not actually a solution, rather a source of error.

**Reply by vferraz** [reply]
I second @fesaaved
